tachyon/math/base/semigroups.h

```diff
@@ -147,12 +157,110 @@ class MultiplicativeSemigroup {
     }
     return g;
   }
+
+  constexpr static std::vector<G> GetSuccessivePowers(size_t size,
+                                                      const G& generator) {
+#if defined(TACHYON_HAS_OPENMP)
+    uint32_t log_size = size_t{base::bits::SafeLog2Ceiling(size)};
+    if (log_size <= kMinLogSizeForParallelization)
+#endif
+      return GetSuccessivePowersSerial(size, generator);
+#if defined(TACHYON_HAS_OPENMP)
+    G power = generator;
+    // [g, g², g⁴, g⁸, ..., g^(2^(|log_size|))]
+    std::vector<G> log_powers = base::CreateVector(log_size, [&power]() {
+      G old_value = power;
+      power.SquareInPlace();
+      return old_value;
+    });
+
+    // allocate the return array and start the recursion
+    std::vector<G> powers =
+        base::CreateVector(size_t{1} << log_size, G::Zero());
+    GetSuccessivePowersRecursive(powers, absl::MakeConstSpan(log_powers));
+    return powers;
+#endif
+  }
+
+ private:
+#if defined(TACHYON_HAS_OPENMP)
+  constexpr static void GetSuccessivePowersRecursive(
+      std::vector<G>& out, const absl::Span<const G>& log_powers) {
+    CHECK_EQ(out.size(), size_t{1} << log_powers.size());
+    // base case: just compute the powers sequentially,
```

Comment:
```suggestion

    // base case: just compute the powers sequentially,
```
Can you add an empty line above?